File: Raven.Abstractions/Data/QueryResult.cs
//-----------------------------------------------------------------------
// <copyright file="QueryResult.cs" company="Hibernating Rhinos LTD">
//     Copyright (c) Hibernating Rhinos LTD. All rights reserved.
// </copyright>
//-----------------------------------------------------------------------
using System;
using System.Collections.Generic;
using Raven.Json.Linq;
using System.Linq;

namespace Raven.Abstractions.Data
{
	/// <summary>
	/// The result of a query
	/// </summary>
	public class QueryResult
	{
		/// <summary>
		/// Gets or sets the document resulting from this query.
		/// </summary>
		/// <value>The results.</value>
		public List<RavenJObject> Results { get; set; }
		/// <summary>
		/// Gets or sets the document included in the result.
		/// </summary>
		/// <value>The includes.</value>
		public List<RavenJObject> Includes { get; set; }

		/// <summary>
		/// Gets or sets a value indicating whether the index is stale.
		/// </summary>
		/// <value><c>true</c> if the index is stale; otherwise, <c>false</c>.</value>
		public bool IsStale { get; set; }

		/// <summary>
		/// The last time the index was updated.
		/// This can be used to determine the freshness of the data.
		/// </summary>
		public DateTime IndexTimestamp { get; set; }

		/// <summary>
		/// Gets or sets the total results for this query
		/// </summary>
		/// <value>The total results.</value>
		public int TotalResults { get; set; }

		/// <summary>
		/// Gets or sets the skipped results
		/// </summary>
		/// <value>The skipped results.</value>
		public int SkippedResults { get; set; }

		/// <summary>
		/// The index used to answer this query
		/// </summary>
		public string IndexName { get; set; }

		/// <summary>
		/// The last etag indexed by the index.
		/// This can be used to determine whatever the results can be cached.
		/// </summary>
		public Etag IndexEtag { get; set; }

		/// <summary>
		/// The ETag value for this index current state, which include what we docs we indexed,
		/// what document were deleted, etc.
		/// </summary>
		public Etag ResultEtag { get; set; }

        /// <summary>
        /// Gets or sets highlighter results 
        /// </summary>
        public Dictionary<string, Dictionary<string, string[]>> Highlightings { get; set; }

		/// <summary>
		/// Gets or sets a value indicating whether any of the documents returned by this query
		/// are non authoritative (modified by uncommitted transaction).
		/// </summary>
		public bool NonAuthoritativeInformation { get; set; }

		/// <summary>
		/// The timestamp of the last time the index was queried
		/// </summary>
		public DateTime LastQueryTime { get; set; }

		/// <summary>
		/// The duration of actually executing the query server side
		/// </summary>
		public long DurationMilliseconds { get; set; }

		/// <summary>
		/// Gets or sets explanations of document scores 
		/// </summary>
		public Dictionary<string, string> ScoreExplanations { get; set; }

		/// <summary>
		/// Initializes a new instance of the <see cref="QueryResult"/> class.
		/// </summary>
		public QueryResult()
		{
			Results = new List<RavenJObject>();
			Includes = new List<RavenJObject>();
		    Highlightings = new Dictionary<string, Dictionary<string, string[]>>();
		}

		/// <summary>
		/// Ensures that the query results can be used in snapshots
		/// </summary>
		public void EnsureSnapshot()
		{
			foreach (var result in Results)
			{
				result.EnsureCannotBeChangeAndEnableSnapshotting();
			}
			foreach (var result in Includes)
			{
				result.EnsureCannotBeChangeAndEnableSnapshotting();
			}
		}

		/// <summary>
		/// Creates a snapshot of the query results
		/// </summary>
		public QueryResult CreateSnapshot()
		{
			return new QueryResult
			{
				Results = new List<RavenJObject>(this.Results.Select(x => (RavenJObject)x.CreateSnapshot())),
				Includes = new List<RavenJObject>(this.Includes.Select(x => (RavenJObject)x.CreateSnapshot())),
				IndexEtag = this.IndexEtag,
				IndexName = this.IndexName,
				IndexTimestamp = this.IndexTimestamp,
				IsStale = this.IsStale,
				SkippedResults = this.SkippedResults,
				TotalResults = this.TotalResults,
				Highlightings = this.Highlightings.ToDictionary(
					pair => pair.Key,
					x => new Dictionary<string, string[]>(x.Value)),
				ScoreExplanations = this.ScoreExplanations.ToDictionary(x => x.Key, x => x.Value)
			};
		}
	}
}

File: Raven.Client.Java/src/main/java/net/ravendb/abstractions/data/QueryResult.java
package net.ravendb.abstractions.data;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import net.ravendb.abstractions.json.linq.RavenJObject;


public class QueryResult {
  private List<RavenJObject> results;
  private List<RavenJObject> includes;
  private boolean isStale;
  private Date indexTimestamp;
  private int totalResults;
  private int skippedResults;
  private String indexName;
  private Etag indexEtag;
  private Etag resultEtag;
  private Map<String, Map<String , String[]>> highlightings;
  private boolean nonAuthoritativeInformation;
  private Date lastQueryTime;
  private long durationMiliseconds;
  private Map<String, String> scoreExplanations;

  /**
   * Initializes a new instance of the {@link QueryResult} class.
   */
  public QueryResult() {
    results = new ArrayList<>();
    includes = new ArrayList<>();
    highlightings = new HashMap<>();
    scoreExplanations = new HashMap<>();
  }

  public Map<String, String> getScoreExplanations() {
    return scoreExplanations;
  }


  public void setScoreExplanations(Map<String, String> scoreExplanations) {
    this.scoreExplanations = scoreExplanations;
  }

  /**
   * Creates a snapshot of the query results
   * @return
   */
  public QueryResult createSnapshot() {
    QueryResult snapshot = new QueryResult();
    for (RavenJObject obj: results) {
      snapshot.getResults().add(obj.createSnapshot());
    }
    for (RavenJObject obj: includes) {
      snapshot.getIncludes().add(obj.createSnapshot());
    }
    snapshot.setIndexEtag(getIndexEtag());
    snapshot.setIndexName(getIndexName());
    snapshot.setIndexTimestamp(getIndexTimestamp());
    snapshot.setStale(isStale);
    snapshot.setSkippedResults(skippedResults);
    snapshot.setTotalResults(getTotalResults());

    for (Map.Entry<String, Map<String, String[]>> entry: highlightings.entrySet()) {
      snapshot.getHighlightings().put(entry.getKey(), new HashMap<>(entry.getValue()));
    }

    for (Map.Entry<String, String> entry: scoreExplanations.entrySet()) {
      snapshot.getScoreExplanations().put(entry.getKey(), entry.getValue());
    }
    return snapshot;
  }

  /**
   * Ensures that the query results can be used in snapshots
   */
  public void ensureSnapshot() {
    for(RavenJObject result: results) {
      result.ensureCannotBeChangeAndEnableShapshotting();
    }
    for(RavenJObject include: includes) {
      include.ensureCannotBeChangeAndEnableShapshotting();
    }
  }

  public long getDurationMiliseconds() {
    return durationMiliseconds;
  }
  public Map<String, Map<String, String[]>> getHighlightings() {
    return highlightings;
  }
  public Collection<RavenJObject> getIncludes() {
    return includes;
  }
  public Etag getIndexEtag() {
    return indexEtag;
  }
  public String getIndexName() {
    return indexName;
  }
  public Date getIndexTimestamp() {
    return indexTimestamp;
  }
  public Date getLastQueryTime() {
    return lastQueryTime;
  }
  public Etag getResultEtag() {
    return resultEtag;
  }
  public List<RavenJObject> getResults() {
    return results;
  }
  public int getSkippedResults() {
    return skippedResults;
  }
  public int getTotalResults() {
    return totalResults;
  }
  public boolean isNonAuthoritativeInformation() {
    return nonAuthoritativeInformation;
  }
  public boolean isStale() {
    return isStale;
  }
  public void setDurationMiliseconds(long durationMiliseconds) {
    this.durationMiliseconds = durationMiliseconds;
  }
  public void setHighlightings(Map<String, Map<String, String[]>> highlightings) {
    this.highlightings = highlightings;
  }
  public void setIncludes(List<RavenJObject> includes) {
    this.includes = includes;
  }
  public void setIndexEtag(Etag indexEtag) {
    this.indexEtag = indexEtag;
  }
  public void setIndexName(String indexName) {
    this.indexName = indexName;
  }
  public void setIndexTimestamp(Date indexTimestamp) {
    this.indexTimestamp = indexTimestamp;
  }
  public void setLastQueryTime(Date lastQueryTime) {
    this.lastQueryTime = lastQueryTime;
  }
  public void setNonAuthoritativeInformation(boolean nonAuthoritativeInformation) {
    this.nonAuthoritativeInformation = nonAuthoritativeInformation;
  }
  public void setResultEtag(Etag resultEtag) {
    this.resultEtag = resultEtag;
  }
  public void setResults(List<RavenJObject> results) {
    this.results = results;
  }
  public void setSkippedResults(int skippedResults) {
    this.skippedResults = skippedResults;
  }
  public void setStale(boolean isStale) {
    this.isStale = isStale;
  }
  public void setTotalResults(int totalResults) {
    this.totalResults = totalResults;
  }




}
